soco/zonegroupstate.py

```diff
@@ -148,8 +150,88 @@ def poll(self, soco):
             )
             soco = soco._satellite_parent
 
-        zgs = soco.zoneGroupTopology.GetZoneGroupState()["ZoneGroupState"]
-        self.process_payload(payload=zgs, source="poll", source_ip=soco.ip_address)
+        # On large (about 20+ players) systems, GetZoneGroupState() can cause
+        # the target Sonos player to return an HTTP 501 error, raising a
+        # SoCoUPnPException.
+        try:
+            zgs = soco.zoneGroupTopology.GetZoneGroupState()["ZoneGroupState"]
+            self.process_payload(payload=zgs, source="poll", source_ip=soco.ip_address)
+
+        # In the event of failure, we fall back to using a ZGT event to
+        # determine the ZGS. Fallback behaviour can be disabled by setting the
+        # config.ZGT_EVENT_FALLBACK flag to False.
+        except SoCoUPnPException as soco_upnp_exception:
+            _LOG.debug(
+                "Exception (%s) raised on 'GetZoneGroupState()'",
+                soco_upnp_exception,
+            )
+
+            if config.ZGT_EVENT_FALLBACK is False:
+                _LOG.debug("ZGT event fallback disabled (config.ZGT_EVENT_FALLBACK)")
+                raise NotSupportedException(
+                    "'GetZoneGroupState()' call fails on large Sonos systems "
+                    "and event fallback is disabled"
+                ) from soco_upnp_exception
+
+            _LOG.debug("Falling back to using a ZGT event")
+            try:
+                self.update_zgs_by_event(soco)
+            except Exception as soco_exception:
+                raise soco_exception from soco_upnp_exception
+
+    def update_zgs_by_event(self, speaker):
+        """
+        Fall back to updating the ZGS using a ZGT event.
+        Use of the 'events_twisted' module is not currently supported.
+        """
+        if config.EVENTS_MODULE.__name__ == "soco.events":
+            _LOG.debug("Updating ZGS using standard 'events' module")
+            self.update_zgs_by_event_default(speaker)
+
+        if config.EVENTS_MODULE.__name__ == "soco.events_asyncio":
+            _LOG.debug("Updating ZGS using 'events_asyncio' module")
+            # Explicit asyncio event loop control required for Python 3.6
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            loop.run_until_complete(ZoneGroupState.update_zgs_by_event_asyncio(speaker))
+            asyncio.set_event_loop(None)
+            loop.close()
+            # From Python 3.7, we can just use the single statement:
+            # asyncio.run(ZoneGroupState.update_zgs_events_asyncio(speaker))
+
+        if config.EVENTS_MODULE.__name__ == "soco.events_twisted":
```

Comment:
Do you maybe want to create a if-elif-elif-else here, with an exception not only in the last elif (events_twisted), but also in the else case. It would make sure that it would fail loudly if we ever grow yet **another** framework for handling events.

Reply:
Good point. I'll make this change.